### PR TITLE
🏎️🏎️🏎️ test runner improvements 🏎️🏎️🏎️

### PR DESCRIPTION
**Pull request by camsaul**
Follow-on improvements to the new test runner I introduced in #16749. These focus on increasing the usefulness of test failure output and getting things to launch and run a little faster.

###### Code reorganization/cleanup

*  Moved `test-runner` and `cloverage` namespaces to `metabase.test-runner` and `metabase.cloverage-runner`, respectively, to make it clearer that it's our custom code wrapping eftest/cloverage respectively.
*  Moved parallel-test related code into new namespace `metabase.test-runner.parallel`. This is a separate namespace mostly to prevent circular deps
*  Moved the redef of `with-redefs` that prevents it from being used in parallel tests to `metabase.test.redefs` so it can be with its friends.

###### Test runner preecondition enforcement

New util fn `metabase.test-runner.init/assert-tests-are-not-initializing` checks whether test namespaces are currently in the process of loading and makes sure you don't do something crazy like try to init the application DB or run queries or stuff like that as a side-effect of loading a test namespace. This stuff shouldn't happen until we actually start running tests anyway (it's very slow and/or annoying to have your tests/REPL hang at launch because it has to init the web server and app DB because a test namespace loaded them as a side effect). Speeds things up a little bit, and keeps `Initializing :something ...` messages from being interspersed with the `Looking for test namespaces in directory ...` messages.

I added assertions to a handful of things to enforce the constraint and fixed a handful of test namespaces that were being naughty.

###### New JUnit outputter
 
`eftest`'s JUnit output is a pale shadow of the beautiful JUnit output we've come to know and love in previous versions of Metabase, and only did output for failed tests (which means we couldn't look at historic execution time for successful tests). This PR adds a custom JUnit test reporter that prints things similarly to how they were printed in the past. 

The old JUnit outputers (both `eftest` and the Leiningen plugin we used to use) waited until the entire test suite finished to write JUnit output, meaning it would add several seconds to the test suite run time; in this PR I tweaked the JUnit new output code to write each namespace in a background thread in a cached thread pool so we get the JUnit output without the extra overhead of waiting for the test runner to write 400 files.

###### Misc perf improvements

*  Our test runner now ignores `src` directories when looking for tests to run which makes things _slightly_ faster.
* Tweaked `metabase.test-runner` a little bit so it starts a little faster, especially for tiny namespaces that don't have lots of dependencies on other stuff in the MB project.

###### Other

* Removed `.lein-classpath` which probably shouldn't have been checked in in the first place, but now at least is definitely no longer needed.
* Not specifically related to the test runner itself, but I replaced some backticks with `[[tags.like/these]]` in docstrings since CIDER at least can just jump right to the definition when it sees those

### CODE REVIEW

File: test/metabase/test_runner/junit/write.clj [hunk 1]
```diff
@@ -130,7 +131,9 @@
      {:classname  (name (ns-name (:ns result)))
       :name       (name (symbol (:var result)))
       :time       (/ (:duration-ms result) 1000.0)
-      :assertions (:assertion-count result)}
+      :assertions (:assertion-count result)
+      ;; NOCOMMIT
```
Comment (dpsutton) [inline test/metabase/test_runner/junit/write.clj:135]: did you mean for a nocommit to end up committed?

Comment (camsaul) [reply test/metabase/test_runner/junit/write.clj:135]: yeah, I was testing whether CircleCI would show the `system-out` attribute in JUnit output. If it did, I was going to write some fancy code to capture System out and System err and include it in JUnit output. Doesn't look like it does tho. So I'll remove this in a second

File: test/metabase/test_runner.clj [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+  (let [tests (u/profile "Find tests" (find-tests only))]
+    (println "Running" (count tests) "tests")
+    tests))
+
+;;;; Running tests & reporting the output
+
+(defonce ^:private orig-test-var t/test-var)
+
+(defn run-test
+  "Run a single test `varr`. Wraps/replaces [[clojure.test/test-var]]."
```
Comment (jeff303) [inline test/metabase/test_runner.clj:98]: Nit: docstring has `varr`, but should it be `test-var`?

File: test/metabase/test_runner/junit.clj [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+(defn handle-event! [{test-var :var, :as event}]
+  (let [test-var (or test-var
+                     (when (seq t/*testing-vars*)
+                       (last t/*testing-vars*)))
+        event    (merge
+                  {:var test-var}
+                  event
+                  (when test-var
+                    {:ns (:ns (meta test-var))}))]
```
Comment (jeff303) [inline test/metabase/test_runner/junit.clj:17]: (for my own edification), is this equivalent to `select-keys`?

**Review by jeff303 — APPROVED** [review]

File: test/metabase/test_runner.clj (hunk 2 above)
Comment (camsaul) [reply test/metabase/test_runner.clj:98]: Good catch

File: test/metabase/test_runner/junit.clj (hunk 3 above)
Comment (camsaul) [reply test/metabase/test_runner/junit.clj:17]: Close but not exactly equivalent. `(select-keys (meta test-var) [:ns])` would only select the key if `:ns` is present:

```clj
(select-keys {:some-other-key 1} [:ns])
=> {}

{:ns (:ns {:some-other-key 1})}
=> {:ns nil}
```

In this case I don't think one form is better than the other -- either works here.

Every `test-var` should have `:ns` metadata -- the only way a var doesn't end up with `:ns` metadata I think is if you created one by hand (i.e., you're doing weird stuff) whereas all the test vars we're running tests for are created by `deftest` and have `:ns` metadata. So I think the form I used works for now

**Comment by camsaul** [comment]
CodeCov is a false positive in this case since the coverage stuff is instrumenting our test runner, which we don't use when running Cloverage. I'll address that in a separate PR (it's only supposed to be instrumenting source files)

**Pull request merged**